File: clients/go/model/player_view.go
package model

import "fmt"
import "io"
import . "komus24/stream"

// Current game's state
type PlayerView struct {
    // Current tick number
    CurrentTick int32
    // TODO - Document
    You Player
    // TODO - Document
    Other []Player
    // TODO - Document
    Quests []Quest
}

func NewPlayerView(currentTick int32, you Player, other []Player, quests []Quest) PlayerView {
    return PlayerView {
        CurrentTick: currentTick,
        You: you,
        Other: other,
        Quests: quests,
    }
}

// Read PlayerView from reader
func ReadPlayerView(reader io.Reader) PlayerView {
    var currentTick int32
    currentTick = ReadInt32(reader)
    var you Player
    you = ReadPlayer(reader)
    var other []Player
    other = make([]Player, ReadInt32(reader))
    for otherIndex := range other {
        var otherElement Player
        otherElement = ReadPlayer(reader)
        other[otherIndex] = otherElement
    }
    var quests []Quest
    quests = make([]Quest, ReadInt32(reader))
    for questsIndex := range quests {
        var questsElement Quest
        questsElement = ReadQuest(reader)
        quests[questsIndex] = questsElement
    }
    return PlayerView {
        CurrentTick: currentTick,
        You: you,
        Other: other,
        Quests: quests,
    }
}

// Write PlayerView to writer
func (playerView PlayerView) Write(writer io.Writer) {
    currentTick := playerView.CurrentTick
    WriteInt32(writer, currentTick)
    you := playerView.You
    you.Write(writer)
    other := playerView.Other
    WriteInt32(writer, int32(len(other)))
    for _, otherElement := range other {
        otherElement.Write(writer)
    }
    quests := playerView.Quests
    WriteInt32(writer, int32(len(quests)))
    for _, questsElement := range quests {
        questsElement.Write(writer)
    }
}

// Get string representation of PlayerView
func (playerView PlayerView) String() string {
    stringResult := "{ "
    stringResult += "CurrentTick: "
    currentTick := playerView.CurrentTick
    stringResult += fmt.Sprint(currentTick)
    stringResult += ", "
    stringResult += "You: "
    you := playerView.You
    stringResult += you.String()
    stringResult += ", "
    stringResult += "Other: "
    other := playerView.Other
    stringResult += "[ "
    for otherIndex, otherElement := range other {
        if otherIndex != 0 {
            stringResult += ", "
        }
        stringResult += otherElement.String()
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "Quests: "
    quests := playerView.Quests
    stringResult += "[ "
    for questsIndex, questsElement := range quests {
        if questsIndex != 0 {
            stringResult += ", "
        }
        stringResult += questsElement.String()
    }
    stringResult += " ]"
    stringResult += " }"
    return stringResult
}

File: clients/csharp/Model/PlayerView.cs
namespace Komus24.Model
{
    /// <summary>
    /// Current game's state
    /// </summary>
    public struct PlayerView
    {
        /// <summary>
        /// Current tick number
        /// </summary>
        public int CurrentTick { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public Komus24.Model.Player You { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public Komus24.Model.Player[] Other { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public Komus24.Model.Quest[] Quests { get; set; }
    
        public PlayerView(int currentTick, Komus24.Model.Player you, Komus24.Model.Player[] other, Komus24.Model.Quest[] quests)
        {
            this.CurrentTick = currentTick;
            this.You = you;
            this.Other = other;
            this.Quests = quests;
        }
    
        /// <summary> Read PlayerView from reader </summary>
        public static PlayerView ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new PlayerView();
            result.CurrentTick = reader.ReadInt32();
            result.You = Komus24.Model.Player.ReadFrom(reader);
            result.Other = new Komus24.Model.Player[reader.ReadInt32()];
            for (int otherIndex = 0; otherIndex < result.Other.Length; otherIndex++)
            {
                result.Other[otherIndex] = Komus24.Model.Player.ReadFrom(reader);
            }
            result.Quests = new Komus24.Model.Quest[reader.ReadInt32()];
            for (int questsIndex = 0; questsIndex < result.Quests.Length; questsIndex++)
            {
                result.Quests[questsIndex] = Komus24.Model.Quest.ReadFrom(reader);
            }
            return result;
        }
    
        /// <summary> Write PlayerView to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(CurrentTick);
            You.WriteTo(writer);
            writer.Write(Other.Length);
            foreach (var otherElement in Other)
            {
                otherElement.WriteTo(writer);
            }
            writer.Write(Quests.Length);
            foreach (var questsElement in Quests)
            {
                questsElement.WriteTo(writer);
            }
        }
    
        /// <summary> Get string representation of PlayerView </summary>
        public override string ToString() {
            string stringResult = "PlayerView { ";
            stringResult += "CurrentTick: ";
            stringResult += CurrentTick.ToString();
            stringResult += ", ";
            stringResult += "You: ";
            stringResult += You.ToString();
            stringResult += ", ";
            stringResult += "Other: ";
            stringResult += "[ ";
            int otherIndex = 0;
            foreach (var otherElement in Other)
            {
                if (otherIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += otherElement.ToString();
                otherIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "Quests: ";
            stringResult += "[ ";
            int questsIndex = 0;
            foreach (var questsElement in Quests)
            {
                if (questsIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += questsElement.ToString();
                questsIndex++;
            }
            stringResult += " ]";
            stringResult += " }";
            return stringResult;
        }
    }
}